src/gui/qgsnewmemorylayerdialog.cpp

```diff
@@ -103,35 +79,40 @@ QgsNewMemoryLayerDialog::~QgsNewMemoryLayerDialog()
 
 QgsWkbTypes::Type QgsNewMemoryLayerDialog::selectedType() const
 {
+  QgsWkbTypes::Type wkbType = QgsWkbTypes::Unknown;
   if ( !buttonGroupGeometry->isChecked() )
   {
-    return QgsWkbTypes::NoGeometry;
+    wkbType = QgsWkbTypes::NoGeometry;
   }
   else if ( mPointRadioButton->isChecked() )
   {
-    return QgsWkbTypes::Point;
+    wkbType = QgsWkbTypes::Point;
   }
   else if ( mLineRadioButton->isChecked() )
   {
-    return QgsWkbTypes::LineString;
+    wkbType = QgsWkbTypes::LineString;
   }
   else if ( mPolygonRadioButton->isChecked() )
   {
-    return QgsWkbTypes::Polygon;
+    wkbType = QgsWkbTypes::Polygon;
   }
   else if ( mMultiPointRadioButton->isChecked() )
   {
-    return QgsWkbTypes::MultiPoint;
+    wkbType = QgsWkbTypes::MultiPoint;
   }
   else if ( mMultiLineRadioButton->isChecked() )
   {
-    return QgsWkbTypes::MultiLineString;
+    wkbType = QgsWkbTypes::MultiLineString;
   }
   else if ( mMultiPolygonRadioButton->isChecked() )
   {
-    return QgsWkbTypes::MultiPolygon;
+    wkbType = QgsWkbTypes::MultiPolygon;
   }
-  return QgsWkbTypes::Unknown;
+
+  if ( mGeometryWithZCheckBox->isChecked() && wkbType != QgsWkbTypes::Unknown && wkbType != QgsWkbTypes::NoGeometry )
```

Comment:
Nicer ui would be to disable (and clear) the mGeometryWithZCheckBox checkbox if the geometry type is set to none by the user.